Fix deconvolution namespace and quality test

elevans opened the pull request:
This PR fixes and refactors the deconvolution namespace :tada:! I compared deconvolution outputs from SciJava Ops with ImageJ Ops to ensure that outputs from both frameworks are the same. This PR also includes an example document with the Groovy and Python script example for Fiji.

A lingering question is whether we should include some pretty output pictures of the deconvolution result like we do in the [PyImageJ docs](https://py.imagej.net/en/latest/Deconvolution.html). I'm a little hesitant to start adding screen shots and other pictures to the repo as it might inflate it. Thoughts?

gselzer committed:
Fix Deconvolution dependencies

elevans committed:
Use OutOfBoundsMirrorFactory for RL decon

Using the OutOfBoundsConstantValueFactory produces
an erroneous deconvolution output. Instead, we
should be using the OutOfBoundsMirrorFactory (as
also done in imagej-ops).

elevans committed:
Remove out of bounds factory for kernel input

Only the input image needs an OutOfBoundsFactory, so there
is no need to have a nullable parameter for an OutOfBoundsFactory
for the kernel.

elevans committed:
Apply obf constant value if non-circulant

If non-circulant mode is desired, apply the
OutOfBoundsConstantValueFactory.

elevans committed:
Update OpRegressionTests

elevans committed:
Encapsulate VectorAccelerator state

- Make VectorAccelerator methods private
- Create acceleration state class

This commit resolves a VectorAccelerator state bug, where
the resources for the VectorAccelerator were being retained
and reused in the next deconvolution run erronously.

This commit creates a new AccelerationState class that stores
these resources and ties their state to the deconvoution Op
(RichardsonLucyC.java).

elevans committed:
Refactor deconvolution namespace

- Create the accelerator Op only in one place (RichardsonLucyC.java).
- Use copy.rai op in VectorAccelerator.
- Add copyright info to AccelerationState.java

elevans committed:
Add Deconvolution example

elevans committed:
Apply maven formatter to deconvolution namespace

gselzer committed:
Move ScriptingInFiji back to the docs directory

gselzer commented on docs/ops/doc/examples/deconvolution.rst:
```diff
@@ -0,0 +1,139 @@
+                                                                pZ,
+                                                                FloatType()).apply()
+
+        # deconvole image
```
deconvole

elevans replied:
Oops! fixed :nerd_face: b73b91bba1cc3456567e6ef77fc70cf16426b83f

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/deconvolve/AccelerationState.java:
```diff
@@ -0,0 +1,49 @@
+import net.imglib2.img.Img;
+import net.imglib2.type.numeric.RealType;
+
+public class AccelerationState<T extends RealType<T>> {
```
Some javadoc would be greatly appreciated

elevans replied:
Done 49cf29dfa0e94c6e397f6cd35ad2f50d6b892981.

gselzer committed:
Simplify nonCirculant conditional

gselzer commented on scijava-ops-image/src/main/java/org/scijava/ops/image/deconvolve/VectorAccelerator.java:
```diff
@@ -56,115 +53,102 @@
  */
-public class VectorAccelerator<T extends RealType<T> & NativeType<T>> implements
-	Inplaces.Arity1<RandomAccessibleInterval<T>>
+public class VectorAccelerator<T extends RealType<T>> implements
```
This Op is interesting now. There's a data structure, *not exposed by the module*, which appears as the input to an Op that users *could* (even though they shouldn't) call. This effectively makes the Op uncallable when JPMS is enforced (note that for now, in the scripts, it is fine). 

@ctrueden @elevans any opinions on this situation? Here are some initial options I see:
* Make this functionality into something that isn't an Op, possibly make it part of the `AcceleratorState`.
* Add a Hint where Ops can declare themselves as hidden.
* Leave it

ctrueden replied:
> There's a data structure, not exposed by the module, which appears as the input to an Op that users could (even though they shouldn't) call.

Arghhhhhhh :sweat_smile: Very good point.

Given the short time frame till the paper, my vote is "Leave it" and file an issue. Any user confusion about it is likely to be near zero—although if someone complains, we can expedite addressing it somehow.

elevans replied:
Thanks @gselzer for explaining the finer details of this issue. My short term vote is to "leave it" as well for the same reason that Curtis has, we just don't have time. Long term, I'm leaning towards demoting the `VectorAccelerator` from an Op to a helper class.

gselzer replied:
In that case let's file an issue. My only concern is that removing an Op is *technically* breaking backwards compatibility, but it's probably fine :wink:

elevans replied:
Hmmm well I'm only leaning that way, so I'm not advocating we remove it right now :nerd_face:. For now we will leave this.

gselzer reviewed (changes requested):
These were all great changes, thank you so much for spearheading them @elevans !

elevans committed:
Fix decon example typos

elevans committed:
Add javadoc to AccelerationState.java

elevans commented:
@gselzer Thanks for the review! I fixed the typos and added the javadoc to `AccelerationState.java`. Lastly, we decided to leave the `VectorAccelerator` as is for now.

elevans committed:
Update decon example title to be more explicit

The previous title just said "deconvolution". The title
should indicate which algorithim we are using.

elevans commented:
Now that we have [media.imagej.net/scijava-ops](https://media.imagej.net/scijava-ops/) I want to add an expected results image to the deconvolution example. So for now, please hold off on merging this PR. I will get this done by tomorrow.

elevans committed:
Improve deconvolution example with images

Added a gif of the RTLV output. And improved
text.

elevans commented:
I updated the decon example. Its ready for a last review. @gselzer, what do you think?

hinerm commented:
@ctrueden @elevans was there a particular reason the `AccelerationState` has pythonic-style field names and/or public parameter access?

Any reason not to a) refactor to getters/setters or b) make the `AccelerationState` protected?

ctrueden commented:
@hinerm The fields are final so it is fine for them to be public.

The `AccelerationState` class needs to be public because it is a parameter type for an Op. Otherwise there would be an Op in the environment that users cannot actually call due to lack of type visibility. Actually, that is already the case due to the package not being exported... but making the class non-public would compound the issue.

ctrueden commented:
Err crap, they *aren't* final! Hmm...

ctrueden commented:
How would you rather proceed?

A) Fix the AccelerationState API and export the package?

OR

B) Make the acceleration thingie not an Op?

I think having private Ops is a bit strange.

OR

C) Decide on and document exactly when Ops should be allowed to be private and why...

hinerm commented:
>A) Fix the AccelerationState API and export the package?

@ctrueden I'm inclined to just do this. I added getters/setters locally and rebuilt. Tests and the script still seem to work.

hinerm committed:
AccelerationState: refactor to getters/setters

Avoids direct access of public fields

hinerm reviewed (approved):
The decon example worked locally on my system, so LGTM.

ctrueden commented:
I am seeing the same test failure on the main branch:
```
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.015 s <<< FAILURE! -- in org.scijava.ops.image.transform.project.project.ProjectTest
  Error:  org.scijava.ops.image.transform.project.project.ProjectTest.testProjector -- Time elapsed: 0.012 s <<< ERROR!
  java.lang.ArrayIndexOutOfBoundsException
```
Is this the race condition with `LoopBuilder`?

ctrueden commented on scijava-ops-image/src/main/java/org/scijava/ops/image/deconvolve/AccelerationState.java:
```diff
@@ -44,15 +44,67 @@
+	}
+
+	public RandomAccessibleInterval<T> ykIterated() {
+		return ykIterated;
```
This field is final, so does not need a getter...
